diff -ur dash-0.5.3.orig/src/Makefile.in dash-0.5.3/src/Makefile.in
--- dash-0.5.3.orig/src/Makefile.in	2005-11-26 05:18:20.000000000 +0200
+++ dash-0.5.3/src/Makefile.in	2007-03-27 00:22:51.000000000 +0300
@@ -57,7 +57,7 @@
 	miscbltin.$(OBJEXT) mystring.$(OBJEXT) options.$(OBJEXT) \
 	parser.$(OBJEXT) redir.$(OBJEXT) show.$(OBJEXT) trap.$(OBJEXT) \
 	output.$(OBJEXT) printf.$(OBJEXT) system.$(OBJEXT) \
-	test.$(OBJEXT) times.$(OBJEXT) var.$(OBJEXT)
+	test.$(OBJEXT) times.$(OBJEXT) var.$(OBJEXT) kdb.$(OBJEXT)
 am_dash_OBJECTS = $(am__objects_1) arith.$(OBJEXT)
 dash_OBJECTS = $(am_dash_OBJECTS)
 dash_DEPENDENCIES = builtins.o init.o nodes.o signames.o syntax.o
@@ -65,7 +65,7 @@
 depcomp = $(SHELL) $(top_srcdir)/depcomp
 am__depfiles_maybe = depfiles
 COMPILE = $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(AM_CPPFLAGS) \
-	$(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS)
+	$(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS) -DDASH
 CCLD = $(CC)
 LINK = $(CCLD) $(AM_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) $(LDFLAGS) -o $@
 YACCCOMPILE = $(YACC) $(YFLAGS) $(AM_YFLAGS)
@@ -88,7 +88,7 @@
 CC = @CC@
 CCDEPMODE = @CCDEPMODE@
 CC_FOR_BUILD = @CC_FOR_BUILD@
-CFLAGS = @CFLAGS@
+CFLAGS = @CFLAGS@ -DDASH
 CPPFLAGS = @CPPFLAGS@
 CYGPATH_W = @CYGPATH_W@
 DEFS = @DEFS@
@@ -167,7 +167,7 @@
 	alias.c arith_yylex.c cd.c error.c eval.c exec.c expand.c \
 	histedit.c input.c jobs.c mail.c main.c memalloc.c miscbltin.c \
 	mystring.c options.c parser.c redir.c show.c trap.c output.c \
-	bltin/printf.c system.c bltin/test.c bltin/times.c var.c
+	bltin/printf.c system.c bltin/test.c bltin/times.c bltin/kdb.c var.c
 
 dash_SOURCES = \
 	$(dash_CFILES) arith.y \
@@ -286,6 +286,7 @@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/system.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/test.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/times.Po@am__quote@
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/kdb.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/trap.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/var.Po@am__quote@
 
@@ -345,6 +346,20 @@
 @AMDEP_TRUE@@am__fastdepCC_FALSE@	DEPDIR=$(DEPDIR) $(CCDEPMODE) $(depcomp) @AMDEPBACKSLASH@
 @am__fastdepCC_FALSE@	$(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS) -c -o times.obj `if test -f 'bltin/times.c'; then $(CYGPATH_W) 'bltin/times.c'; else $(CYGPATH_W) '$(srcdir)/bltin/times.c'; fi`
 
+kdb.o: bltin/kdb.c
+@am__fastdepCC_TRUE@	if $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS) -MT kdb.o -MD -MP -MF "$(DEPDIR)/kdb.Tpo" -c -o kdb.o `test -f 'bltin/kdb.c' || echo '$(srcdir)/'`bltin/kdb.c; \
+@am__fastdepCC_TRUE@	then mv -f "$(DEPDIR)/kdb.Tpo" "$(DEPDIR)/kdb.Po"; else rm -f "$(DEPDIR)/kdb.Tpo"; exit 1; fi
+@AMDEP_TRUE@@am__fastdepCC_FALSE@	source='bltin/kdb.c' object='kdb.o' libtool=no @AMDEPBACKSLASH@
+@AMDEP_TRUE@@am__fastdepCC_FALSE@	DEPDIR=$(DEPDIR) $(CCDEPMODE) $(depcomp) @AMDEPBACKSLASH@
+@am__fastdepCC_FALSE@	$(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS) -c -o kdb.o `test -f 'bltin/kdb.c' || echo '$(srcdir)/'`bltin/kdb.c
+
+kdb.obj: bltin/kdb.c
+@am__fastdepCC_TRUE@	if $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS) -MT kdb.obj -MD -MP -MF "$(DEPDIR)/kdb.Tpo" -c -o kdb.obj `if test -f 'bltin/kdb.c'; then $(CYGPATH_W) 'bltin/kdb.c'; else $(CYGPATH_W) '$(srcdir)/bltin/kdb.c'; fi`; \
+@am__fastdepCC_TRUE@	then mv -f "$(DEPDIR)/kdb.Tpo" "$(DEPDIR)/kdb.Po"; else rm -f "$(DEPDIR)/kdb.Tpo"; exit 1; fi
+@AMDEP_TRUE@@am__fastdepCC_FALSE@	source='bltin/kdb.c' object='kdb.obj' libtool=no @AMDEPBACKSLASH@
+@AMDEP_TRUE@@am__fastdepCC_FALSE@	DEPDIR=$(DEPDIR) $(CCDEPMODE) $(depcomp) @AMDEPBACKSLASH@
+@am__fastdepCC_FALSE@	$(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS) -c -o kdb.obj `if test -f 'bltin/kdb.c'; then $(CYGPATH_W) 'bltin/kdb.c'; else $(CYGPATH_W) '$(srcdir)/bltin/kdb.c'; fi`
+
 .y.c:
 	$(YACCCOMPILE) $<
 	if test -f y.tab.h; then \
diff -ur dash-0.5.3.orig/src/builtins.def.in dash-0.5.3/src/builtins.def.in
--- dash-0.5.3.orig/src/builtins.def.in	2005-11-26 05:17:55.000000000 +0200
+++ dash-0.5.3/src/builtins.def.in	2007-03-27 00:20:26.000000000 +0300
@@ -72,6 +72,7 @@
 jobscmd		-u jobs
 localcmd	-a local
 printfcmd	printf
+kdbcmd	kdb
 pwdcmd		pwd
 readcmd		-u read
 returncmd	-s return
