--- bluez-utils-2.25/hcid/hcid.conf     2000-01-01 02:02:18.000000000 +0200
+++ bluez-utils-2.25-new/hcid/hcid.conf 2005-10-30 02:04:28.000000000 +0300
@@ -1,64 +1,21 @@
 #
 # HCI daemon configuration file.
 #
-
-# HCId options
 options {
-	# Automatically initialize new devices
 	autoinit yes;
-
-	# Security Manager mode
-	#   none - Security manager disabled
-	#   auto - Use local PIN for incoming connections
-	#   user - Always ask user for a PIN
-	#
-	security auto;
-
-	# Pairing mode
-	#   none  - Pairing disabled
-	#   multi - Allow pairing with already paired devices
-	#   once  - Pair once and deny successive attempts
+	security none;
 	pairing multi;
-
-	# PIN helper
-	pin_helper /usr/bin/bluepin;
-
-	# D-Bus PIN helper
-	#dbus_pin_helper;
+	pin_helper /etc/bluetooth/givepin;
 }
-
-# Default settings for HCI devices
 device {
-	# Local device name
-	#   %d - device id
-	#   %h - host name
-	name "BlueZ (%d)";
-
-	# Local device class
+	name "%h-%d";
 	class 0x3e0100;
-
-	# Default packet type
+	#class 0x000100; # computer
 	#pkt_type DH1,DM1,HV1;
-
-	# Inquiry and Page scan
 	iscan enable; pscan enable;
-
-	# Default link mode
-	#   none   - no specific policy 
-	#   accept - always accept incoming connections
-	#   master - become master on incoming connections,
-	#            deny role switch on outgoing connections
-	lm accept;
-
-	# Default link policy
-	#   none    - no specific policy
-	#   rswitch - allow role switch
-	#   hold    - allow hold mode
-	#   sniff   - allow sniff mode
-	#   park    - allow park mode
-	lp rswitch,hold,sniff,park;
-
-	# Authentication and Encryption (Security Mode 3)
+	#discovto 0;
+	lm master;
+	#lp rswitch,hold,sniff,park;
 	#auth enable;
 	#encrypt enable;
 }
