--- linux-2.4.32/arch/mips/am5120/serial.c.orig	2008-03-31 21:53:46.000000000 +0200
+++ linux-2.4.32/arch/mips/am5120/serial.c	2008-03-31 21:58:51.000000000 +0200
@@ -25,6 +25,7 @@
 ;Modification History:
 : 1.second uart 12.12.2005
 ; 2.chahge speed  3.3.2006
+; 3.lcr_h register fixup 31.3.2008 Slawomir Cygan ( slaffek_gmail.com )
 ;
 ;*****************************************************************************/
 #include <linux/config.h>
@@ -127,6 +128,7 @@
 	prom_port_info.magic = SERIAL_MAGIC;
 	prom_port_info.port = ser->port;
 	prom_port_info.flags = ser->flags;
+	prom_port_info.LCR = UART_WLEN_8BITS | UART_ENABLE_FIFO;
 
 	/* set baudrate to 115200 */
 	divisor = ADM5120_UARTCLK_FREQ/(16*(prom_port_info.state->baud_base)) - 1;
@@ -597,6 +599,7 @@
 	serial_out(info, UART_LCR_L_REG, divisor); 
 	serial_out(info, UART_LCR_M_REG, divisor >> 8);
 	serial_out(info, UART_LCR_H_REG, (UART_WLEN_8BITS | UART_ENABLE_FIFO));
+	info->LCR = UART_WLEN_8BITS | UART_ENABLE_FIFO;
 
 	/*
 	 * Finally, enable interrupts
@@ -700,6 +703,7 @@
 	serial_out(info, UART_LCR_L_REG,divisor);
 	serial_out(info, UART_LCR_M_REG,divisor >> 8);
 	serial_out(info, UART_LCR_H_REG, (bits |stops | pars | UART_ENABLE_FIFO));
+	info->LCR = (bits |stops | pars | UART_ENABLE_FIFO);
 	restore_flags(flags);
 
 }
